pkg/pillar/docs/zedkube.md

```diff
@@ -0,0 +1,47 @@
+
+### Cluster Status
+
+The kubevirt EVE image supports either running in kubernetes with single node or with cluster mode (at least 3 nodes). In single-node mode, there is no change to the EVE API from the controller with EVE devices. In the cluster mode, controller will send 'EdgeNodeClusterConfig' to the device, and being published in 'zedagent' and zedkube subscribe to that, then it will publish the 'EdgeNodeClusterStatus'. The container 'kube' will monitor the 'EdgeNodeClusterStatus' for cluster mode changes. zedkube will subscribe also to the deviceNetworkStatus from 'nim' and make sure the cluster prefix is ready for the kubernetes.
+
+### Collect App Container Logs
+
+A timer job is used to search through the AppInstanceConfig subscription list, and if the application is a native container, without a VM shim layer, it searches the pods with the application name, and use the kubernetes pods 'GetLogs' API to acquire the logs since last query. It then makes the log entry with App UUID, container name and timestamp, so the newlogd will pick this up for the app log collection.
+
+### Publish Cluster Application Status
+
+zedkube periodically query all the PODs running on this device and report each of the applications in AppInstanceConfig. The main purpose of this status is for application migration handling. We may have the App which sets the Designated Node ID to itself, and the application may moved onto another node; or we are the node which does not match the App's Designated Node ID, but it appears to be scheduled on this node. The 'zedmanager' listens on this ENClusterAppStatus, and makes the decision for the 'effective' activate flag of the application.
+
+In VMI case, there will always be a POD having the name with prefix 'virt-launcher-', we only use this to determine the migration status of the VMI.
+
+### Cluster Status Server
+
+In the Kubernetes cluster mode with multiple HA servers, when it starts up, it needs to join the cluster by specifying the 'bootstrap' server IP address. Even if the IP address is there, sometimes the 'bootstrap' node is still in single-node mode or it has not been converted into the cluster server yet. This will create problem for the joining server, and will later have conflicts with the status and certificates or tokens. To handle this server joining, zedkube is responsible for reporting it's cluster status through HTTP service. Each of the cluster servers will have a HTTP service on the cluster interface with port number '12346' using URL /status. It will report status of 'cluster' if the node has the property of 'master' and 'etcd'. The new joining server node or agent node will not move forward for starting the kubernetes node unless the http query returns 'cluster' status over the cluster network. The 'ClusterStatus' port for the HTTP is explicitly allowed on EVE firewall.
```

Comment:
Are we restricting this to use the cluster network interface and IP addresses using some ACLs? Can we use https using some certificates which k3s might have?

I don't think we can have TCP port number 12346 accessible from the Internet.

Reply:
Currently this is just HTTP, using certs only going to very the server side which does not help. Yes, this is only listening on the cluster IP which is a private prefix setup locally.

Reply:
the firewall check is specific to the cluster interface and cluster IP w/ this port:
```
		clusterStatusRule := iptables.Rule{
			RuleLabel: "Allow access to Cluster Status",
			MatchOpts: []string{"-p", "tcp", "-i", clusterPort.IfName,
				"-d", clusterStatus.ClusterIPPrefix.IP.String(), "--dport", "12346"},
			Target:      "ACCEPT",
			Description: "Allow access to Cluster Status via cluster interface",
		}
```